# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
pisa.dtypes[400:]

ST23Q08        object
ST24Q01        object
ST24Q02        object
ST24Q03        object
CLCUSE1        object
CLCUSE301       int64
CLCUSE302       int64
DEFFORT         int64
QUESTID        object
BOOKID         object
EASY           object
AGE           float64
GRADE         float64
PROGN          object
ANXMAT        float64
ATSCHL        float64
ATTLNACT      float64
BELONG        float64
BFMJ2         float64
BMMJ1         float64
CLSMAN        float64
COBN_F         object
COBN_M         object
COBN_S         object
COGACT        float64
CULTDIST      float64
CULTPOS       float64
DISCLIMA      float64
ENTUSE        float64
ESCS          float64
               ...   
W_FSTR55      float64
W_FSTR56      float64
W_FSTR57      float64
W_FSTR58      float64
W_FSTR59      float64
W_FSTR60      float64
W_FSTR61      float64
W_FSTR62      float64
W_FSTR63      float64
W_FSTR64      float64
W_FSTR65      float64
W_FSTR66      float64
W_FSTR67      float64
W_FSTR68      float64
W_FSTR69  

In [6]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [7]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [8]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [9]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [10]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [11]:
pisa_clean = pisa_clean[['STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [12]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [13]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score']) / 3

In [14]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [15]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        472.488764
std          96.036271
min          77.114593
25%         403.992595
50%         472.046460
75%         541.455700
max         826.592027
Name: Average Total Score, dtype: float64


In [16]:
pisa_clean.sample(10)

STIDSTD ST04Q01                ST13Q01 ST14Q01 ST14Q02 ST14Q03  \
382588     3667  Female      <ISCED level 3A>       No     Yes      No   
397273     8023    Male  <ISCED level 3B, 3C>       No      No      No   
457566     3190    Male       <ISCED level 1>       No      No     NaN   
157894    12312    Male      <ISCED level 3A>       No      No     Yes   
229800     2790    Male      <ISCED level 3A>       No      No      No   
160452    14870  Female                    NaN      No      No      No   
102029     6016  Female  <ISCED level 3B, 3C>       No      No      No   
404338     4122    Male      <ISCED level 3A>       No     Yes      No   
282917     4217    Male  <ISCED level 3B, 3C>       No      No     NaN   
40404      3773    Male  <ISCED level 3B, 3C>       No      No      No   

       ST14Q04                ST17Q01 ST18Q01 ST18Q02  ... ST57Q03 ST57Q04  \
382588      No  <ISCED level 3B, 3C>       No      No  ...     NaN     NaN   
397273     Yes  <ISCED level 3B, 3C>       No     NaN  ...     NaN     NaN   
457566      No       <ISCED level 1>       No      No  ...     1.0     1.0   
157894      No      <ISCED level 3A>       No     Yes  ...     NaN     NaN   
229800     Yes      <ISCED level 3A>       No      No  ...     1.0     1.0   
160452      No      <ISCED level 3A>       No     Yes  ...     NaN     NaN   
102029      No  <ISCED level 3B, 3C>       No      No  ...     0.0     0.0   
404338      No      <ISCED level 3A>       No     Yes  ...     NaN     NaN   
282917     NaN      <ISCED level 3A>       No     Yes  ...     0.0     0.0   
40404       No  <ISCED level 3B, 3C>       No      No  ...     NaN     NaN   

        ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
382588      NaN    NaN    NaN    NaN           571.90372   
397273      NaN    NaN    NaN    NaN           360.11062   
457566      1.0  120.0  180.0    NaN           294.44618   
157894      NaN    NaN    NaN    NaN           406.92474   
229800      0.0  200.0  200.0  160.0           613.18742   
160452      NaN    NaN    NaN    NaN           316.02274   
102029      1.0  225.0  225.0  135.0           583.97726   
404338      NaN    NaN    NaN    NaN           662.10470   
282917      0.0  250.0  200.0  200.0           713.98194   
40404       NaN    NaN    NaN    NaN           612.64216   

        Average Reading Score  Average Science Score  Average Total Score  
382588              533.54204              574.07682           559.840860  
397273              387.16082              412.19696           386.489467  
457566              306.16438              351.11900           317.243187  
157894              341.28958              393.17418           380.462833  
229800              615.87556              640.28348           623.115487  
160452              214.22940              287.70982           272.653987  
102029              583.10700              570.25364           579.112633  
404338              597.51102              648.76910           636.128273  
282917              696.07006              762.34614           724.132713  
40404               590.61430              620.23498           607.830480  

[10 rows x 24 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [18]:
# Rename columns
pisa_clean.rename({'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning Time - Mathematics',
                   'LMINS': 'Learning Time - Test Language',
                   'SMINS': 'Learning Time - Science'}, axis='columns', inplace=True)

In [19]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

# Null

In [23]:
# pisa_clean.replace(["NaN"], np.nan, inplace = True)
# pisa_clean.dropna(inplace = True)

In [24]:
pisa_clean.duplicated().sum()

0

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [25]:
father_edu = pisa_clean.copy()

In [26]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [27]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [28]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [29]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [30]:
father_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
385437        1909  Female                  <ISCED level 3B, 3C>    
357414        3162  Female                       <ISCED level 1>    
91875        17406    Male                      <ISCED level 3A>    
142891        4790  Female                      <ISCED level 3A>    
461026          44  Female  She did not complete <ISCED level 1>    
159443       13861  Female                       <ISCED level 1>    
110471        3229  Female                      <ISCED level 3A>    
106888       10875    Male                      <ISCED level 3A>    
313103        2736  Female                       <ISCED level 1>    
163978       18396    Male                      <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
385437                                    No   
357414                                    No   
91875                                     No   
142891                                    No   
461026                                    No   
159443                                    No   
110471                                    No   
106888                                   Yes   
313103                                    No   
163978                                    No   

       Mother Qualifications - ISCED level 5A  \
385437                                    Yes   
357414                                     No   
91875                                      No   
142891                                     No   
461026                                     No   
159443                                     No   
110471                                    Yes   
106888                                    Yes   
313103                                     No   
163978                                     No   

       Mother Qualifications - ISCED level 5B  \
385437                                     No   
357414                                     No   
91875                                      No   
142891                                    Yes   
461026                                     No   
159443                                     No   
110471                                     No   
106888                                     No   
313103                                     No   
163978                                    Yes   

       Mother Qualifications - ISCED level 4  \
385437                                    No   
357414                                    No   
91875                                     No   
142891                                   Yes   
461026                                    No   
159443                                    No   
110471                                    No   
106888                                   Yes   
313103                                    No   
163978                                    No   

                  Father - Highest Schooling <ISCED level 6>  <ISCED level 5>  \
385437                      <ISCED level 2>               No               No   
357414                     <ISCED level 3A>               No               No   
91875                      <ISCED level 3A>               No               No   
142891                 <ISCED level 3B, 3C>               No               No   
461026  He did not complete <ISCED level 1>               No               No   
159443                      <ISCED level 1>               No               No   
110471                     <ISCED level 3A>               No  <ISCED level 5>   
106888                 <ISCED level 3B, 3C>               No               No   
313103                      <ISCED level 1>               No               No   
163978                      <ISCED level 2>               No               No   

        ... Learning Time - Science Average Math Score  Average Reading Score  \
385437  ...                   180.0          329.57624              406.61128   
357414  ...                   240.0          401.86166  

In [31]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [32]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
250487        13385    Male           <ISCED level 2>    
686811         1433    Male          <ISCED level 3A>    
849081         1501    Male          <ISCED level 3A>    
425841          770  Female           <ISCED level 2>    
518319         3832  Female          <ISCED level 3A>    
1035837        1596  Female           <ISCED level 2>    
748307          312  Female          <ISCED level 3A>    
257520        10634  Female          <ISCED level 3A>    
67658          3659    Male           <ISCED level 2>    
1012353        3704    Male           <ISCED level 2>    

        Mother Qualifications - ISCED level 6  \
250487                                     No   
686811                                     No   
849081                                     No   
425841                                     No   
518319                                     No   
1035837                                    No   
748307                                    Yes   
257520                                     No   
67658                                      No   
1012353                                    No   

        Mother Qualifications - ISCED level 5A  \
250487                                      No   
686811                                      No   
849081                                      No   
425841                                      No   
518319                                     Yes   
1035837                                     No   
748307                                     Yes   
257520                                     Yes   
67658                                       No   
1012353                                     No   

        Mother Qualifications - ISCED level 5B  \
250487                                      No   
686811                                      No   
849081                                      No   
425841                                      No   
518319                                     Yes   
1035837                                     No   
748307                                      No   
257520                                      No   
67658                                       No   
1012353                                     No   

        Mother Qualifications - ISCED level 4  \
250487                                     No   
686811                                     No   
849081                                    Yes   
425841                                    Yes   
518319                                     No   
1035837                                   Yes   
748307                                     No   
257520                                     No   
67658                                      No   
1012353                                    No   

         Out-of-School Study Time - Homework  \
250487                                   4.0   
686811                                   1.0   
849081                                   5.0   
425841                                   6.0   
518319                                   8.0   
1035837                                  6.0   
748307                                  15.0   
257520                                   3.0   
67658                                    1.0   
1012353                                  1.0   

         Out-of-School Study Time - Guided Homework  \
250487                                          3.0   
686811                                          0.0   
849081                                          2.0   
425841                                          3.0   
518319                                          0.0   
1035837                                         1.0   
748307                                          1.0   
257520                                          0.0   
67658                                           1.0   
1012353                                         1.0   

         Out-of-School Study Time - Personal Tutor  .

In [33]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [34]:
father_edu.sample(70)

Student ID  Gender Mother - Highest Schooling  \
102937         1651    Male          <ISCED level 3A>    
90771          1497    Male      <ISCED level 3B, 3C>    
867187         1777  Female          <ISCED level 3A>    
6563          12635  Female          <ISCED level 3A>    
658944         3319    Male           <ISCED level 1>    
107356         2681    Male           <ISCED level 2>    
1953           4945    Male          <ISCED level 3A>    
787143         1858  Female           <ISCED level 2>    
755906         3947    Male           <ISCED level 2>    
48001          1222  Female      <ISCED level 3B, 3C>    
987794        10484  Female           <ISCED level 2>    
113006          413    Male          <ISCED level 3A>    
34966          1423    Male          <ISCED level 3A>    
17858          4030  Female          <ISCED level 3A>    
32677          4125    Male          <ISCED level 3A>    
290015         1409    Male          <ISCED level 3A>    
600240         7286  Female      <ISCED level 3B, 3C>    
77592         29812    Male           <ISCED level 2>    
263789         4980    Male          <ISCED level 3A>    
806            4103  Female          <ISCED level 3A>    
33669          1895  Female           <ISCED level 2>    
392205         1625  Female          <ISCED level 3A>    
749941         4305  Female      <ISCED level 3B, 3C>    
57864          4593  Female           <ISCED level 2>    
79773             5  Female      <ISCED level 3B, 3C>    
45484         24737    Male           <ISCED level 2>    
976490         2508    Male          <ISCED level 3A>    
904001          483    Male          <ISCED level 3A>    
100439         4928    Male           <ISCED level 1>    
38881          4671    Male          <ISCED level 3A>    
...             ...     ...                        ...   
698853         6581  Female      <ISCED level 3B, 3C>    
143587         2323  Female          <ISCED level 3A>    
175756         2470  Female          <ISCED level 3A>    
566427         3443  Female          <ISCED level 3A>    
820983         4961  Female          <ISCED level 3A>    
92916           505    Male      <ISCED level 3B, 3C>    
66884          1577  Female           <ISCED level 2>    
829891         7984  Female          <ISCED level 3A>    
1026249         471    Male          <ISCED level 3A>    
261920         3813    Male          <ISCED level 3A>    
434038         3869  Female          <ISCED level 3A>    
876985        20771  Female          <ISCED level 3A>    
705908         3823    Male          <ISCED level 3A>    
966862         4900    Male      <ISCED level 3B, 3C>    
42550         15838    Male          <ISCED level 3A>    
949643         2913    Male          <ISCED level 3A>    
53723          8408    Male          <ISCED level 3A>    
345963         4711  Female          <ISCED level 3A>    
681021         3227    Male          <ISCED level 3A>    
14058          6945    Male           <ISCED level 1>    
587245         5683    Male           <ISCED level 1>    
6941          14050  Female           <ISCED level 2>    
64107          1009  Female           <ISCED level 2>    
355729         1145    Male          <ISCED level 3A>    
67620          3526    Male      <ISCED level 3B, 3C>    
53038          6591  Female          <ISCED level 3A>    
80063           489  Female      <ISCED level 3B, 3C>    
46515          2139  Female          <ISCED level 3A>    
432382         3108    Male          <ISCED level 3A>    
66320           125    Male          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
102937                                     No   
90771                                      No   
867187                                    Yes   
6563                                       No   
658944                                     No   
107356                                     No   
1953                                       No   
787143              

In [35]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [36]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    84496
<ISCED level 5>    49085
<ISCED level 4>    26411
<ISCED level 2>    21955
<ISCED level 6>     6715
<ISCED level 1>     5947
<ISCED level 0>     2290
Name: Education - Father, dtype: int64

In [37]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [38]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
961480        20263    Male      <ISCED level 3B, 3C>    
891622           26  Female          <ISCED level 3A>    
251798        17102    Male          <ISCED level 3A>    
1006741        1400  Female      <ISCED level 3B, 3C>    
531380        16676  Female           <ISCED level 2>    
515839         2840  Female          <ISCED level 3A>    
776311         2665    Male          <ISCED level 3A>    
861112          901    Male          <ISCED level 3A>    
238418          755  Female          <ISCED level 3A>    
292865         4232  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
961480                                     No   
891622                                     No   
251798                                     No   
1006741                                    No   
531380                                     No   
515839                                     No   
776311                                     No   
861112                                     No   
238418                                     No   
292865                                     No   

        Mother Qualifications - ISCED level 5A  \
961480                                      No   
891622                                      No   
251798                                     Yes   
1006741                                     No   
531380                                      No   
515839                                      No   
776311                                      No   
861112                                      No   
238418                                     Yes   
292865                                     Yes   

        Mother Qualifications - ISCED level 5B  \
961480                                      No   
891622                                      No   
251798                                     Yes   
1006741                                     No   
531380                                      No   
515839                                     Yes   
776311                                      No   
861112                                      No   
238418                                      No   
292865                                      No   

        Mother Qualifications - ISCED level 4  \
961480                                    Yes   
891622                                    Yes   
251798                                     No   
1006741                                    No   
531380                                     No   
515839                                    Yes   
776311                                    Yes   
861112                                     No   
238418                                     No   
292865                                     No   

         Out-of-School Study Time - Homework  \
961480                                  10.0   
891622                                   3.0   
251798                                  10.0   
1006741                                  2.0   
531380                                  30.0   
515839                                  18.0   
776311                                   5.0   
861112                                  15.0   
238418                                  14.0   
292865                                  12.0   

         Out-of-School Study Time - Guided Homework  \
961480                                          9.0   
891622                                          2.0   
251798                                          2.0   
1006741                                         0.0   
531380                                          0.0   
515839                                          0.0   
776311                                          0.0   
861112                                          3.0   
238418                                          0.0   
292865                                          0.0   

         Out-of-School Study Time - Personal Tutor  \

In [39]:
father_edu.duplicated().sum()

6947

In [40]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')

father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)
father_edu_clean = father_edu.drop_duplicates(subset='Student ID', keep='last')

In [41]:
father_edu_clean.duplicated().sum()

0

In [42]:
father_edu.shape

(196899, 20)

### Mother Edu

In [43]:
mother_edu = father_edu_clean.copy()

In [44]:
mother_edu.head()

Student ID  Gender             Mother - Highest Schooling  \
1062054        6301  Female  She did not complete <ISCED level 1>    
1054542       16914    Male                      <ISCED level 3A>    
1054388       16487    Male                      <ISCED level 3A>    
1054076       15611  Female                      <ISCED level 3A>    
1053584       14175    Male                      <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
1062054                                    No   
1054542                                    No   
1054388                                    No   
1054076                                    No   
1053584                                    No   

        Mother Qualifications - ISCED level 5A  \
1062054                                     No   
1054542                                    Yes   
1054388                                     No   
1054076                                     No   
1053584                                    Yes   

        Mother Qualifications - ISCED level 5B  \
1062054                                     No   
1054542                                     No   
1054388                                     No   
1054076                                     No   
1053584                                     No   

        Mother Qualifications - ISCED level 4  \
1062054                                    No   
1054542                                    No   
1054388                                    No   
1054076                                    No   
1053584                                    No   

         Out-of-School Study Time - Homework  \
1062054                                  0.0   
1054542                                  0.0   
1054388                                  1.0   
1054076                                  2.0   
1053584                                  1.0   

         Out-of-School Study Time - Guided Homework  \
1062054                                         0.0   
1054542                                         0.0   
1054388                                         1.0   
1054076                                         0.0   
1053584                                         1.0   

         Out-of-School Study Time - Personal Tutor  \
1062054                                        0.0   
1054542                                        0.0   
1054388                                        0.0   
1054076                                        0.0   
1053584                                        0.0   

         Out-of-School Study Time - Commercial Company  \
1062054                                            0.0   
1054542                                            0.0   
1054388                                            0.0   
1054076                                            0.0   
1053584                                            0.0   

         Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
1062054                                     1.0                        540.0   
1054542                                     0.0                        280.0   
1054388                                     1.0                        400.0   
1054076                                     0.0                        375.0   
1053584                                     1.0                        400.0   

         Learning Time - Test Language  Learning Time - Science  \
1062054                          720.0                     90.0   
1054542                          280.0                    280.0   
1054388                          400.0                    400.0   
1054076                          375.0                    375.0   
1053584                            0.0                      0.0   

         Average Math Score  Average Reading Score  Average Science Score  \
1062054           301.69028              311.29408              287.52330   
1054542           497.43728              474.25206              

In [45]:
mother_edu['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         15483
<ISCED level 3B, 3C>                      3806
<ISCED level 2>                           3612
<ISCED level 1>                            533
She did not complete <ISCED level 1>       151
Name: Mother - Highest Schooling, dtype: int64

In [46]:
pisa_clean['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         63660
<ISCED level 3B, 3C>                     23596
<ISCED level 2>                          19860
<ISCED level 1>                           5189
She did not complete <ISCED level 1>      2383
Name: Mother - Highest Schooling, dtype: int64

In [47]:
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [48]:
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [49]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [50]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [51]:
mother_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  <ISCED level 6>  \
877498       22114    Male      <ISCED level 3B, 3C>                No   
500015       11938    Male          <ISCED level 3A>                No   
266580        7005  Female          <ISCED level 3A>                No   
310309        2921  Female          <ISCED level 3A>                No   
274260       22940  Female          <ISCED level 3A>                No   
248859        8656  Female          <ISCED level 3A>                No   
874148       13242    Male          <ISCED level 3A>                No   
308295        2971  Female          <ISCED level 3A>   <ISCED level 6>   
535758       28220  Female           <ISCED level 2>                No   
729969       13752    Male           <ISCED level 2>                No   

        <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
877498               No               No               No   
500015               No               No               No   
266580               No  <ISCED level 5>               No   
310309  <ISCED level 5>  <ISCED level 5>  <ISCED level 4>   
274260  <ISCED level 5>               No               No   
248859  <ISCED level 5>  <ISCED level 5>               No   
874148  <ISCED level 5>               No               No   
308295  <ISCED level 5>               No               No   
535758               No               No  <ISCED level 4>   
729969               No               No               No   

        Out-of-School Study Time - Homework  \
877498                                  5.0   
500015                                  3.0   
266580                                 10.0   
310309                                  8.0   
274260                                  5.0   
248859                                 11.0   
874148                                 12.0   
308295                                 20.0   
535758                                 20.0   
729969                                  5.0   

        Out-of-School Study Time - Guided Homework  \
877498                                         0.0   
500015                                         3.0   
266580                                         0.0   
310309                                         0.0   
274260                                         3.0   
248859                                         2.0   
874148                                         0.0   
308295                                        14.0   
535758                                         1.0   
729969                                         0.0   

        Out-of-School Study Time - Personal Tutor  ...  Average Math Score  \
877498                                        0.0  ...           554.37764   
500015                                        2.0  ...           573.61736   
266580                                        0.0  ...           583.89936   
310309                                        0.0  ...           588.65086   
274260                                        4.0  ...           519.24754   
248859                                        0.0  ...           490.19316   
874148                                        0.0  ...           646.52596   
308295                                       10.0  ...           479.13228   
535758                                        1.0  ...           565.98382   
729969                                        0.0  ...           532.02212   

        Average Reading Score  Average Science Score  Average Total Score  \
877498              537.84630              587.59790           559.940613   
500015              479.22410              537.98958           530.277013   
266580              553.08206              590.30212           575.761180   
310309              567.06192              614.26706           589.993280   
274260              536.71930              530.06342           528.676753   
248859              415.03100              452.01418           452.412780   
874148    

In [52]:
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning Time - Mathematics',
                            'Learning Time - Test Language',
                            'Learning Time - Science',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [53]:
mother_edu.sample(10)

Student ID  Gender  Out-of-School Study Time - Homework  \
185093        6603  Female                                  7.0   
178116        6551    Male                                  2.0   
185288        1881    Male                                  4.0   
75110        30862  Female                                 10.0   
226428       14635  Female                                  8.0   
55112        12233    Male                                  2.0   
221992       17264    Male                                  1.0   
1155         23281  Female                                 16.0   
90314         3252  Female                                  3.0   
226546       13351    Male                                 10.0   

        Out-of-School Study Time - Guided Homework  \
185093                                         5.0   
178116                                         0.0   
185288                                         0.0   
75110                                          0.0   
226428                                         0.0   
55112                                          0.0   
221992                                         0.0   
1155                                           4.0   
90314                                          1.0   
226546                                         5.0   

        Out-of-School Study Time - Personal Tutor  \
185093                                        3.0   
178116                                        0.0   
185288                                        0.0   
75110                                         0.0   
226428                                        0.0   
55112                                         0.0   
221992                                        0.0   
1155                                          4.0   
90314                                         0.0   
226546                                        0.0   

        Out-of-School Study Time - Commercial Company  \
185093                                            0.0   
178116                                            0.0   
185288                                            0.0   
75110                                             0.0   
226428                                            0.0   
55112                                             0.0   
221992                                            0.0   
1155                                              0.0   
90314                                             2.0   
226546                                            0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
185093                                     1.0                        180.0   
178116                                     0.0                        250.0   
185288                                     0.0                        240.0   
75110                                      0.0                        150.0   
226428                                     0.0                        200.0   
55112                                      0.0                        210.0   
221992                                     1.0                        300.0   
1155                                       0.0                        300.0   
90314                                      1.0                        225.0   
226546                                     7.0                        240.0   

        Learning Time - Test Language  Learning Time - Science  \
185093                          180.0                    180.0   
178116                          350.0                    400.0   
185288                          240.0                    180.0   
75110                           200.0                    100.0   
226428                          200.0                    100.0   
55112                           210.0                    210.0   
221992                          300.0                    300.0   
1155                            240.0                    120.0   

In [54]:
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [55]:
mother_edu.sample(70)

Student ID  Gender  Out-of-School Study Time - Homework  \
8657         12522  Female                                 10.0   
3916         28843    Male                                  7.0   
16884        16172    Male                                  3.0   
65416         9333  Female                                  5.0   
186567       10162  Female                                  6.0   
176798       21445    Male                                  4.0   
114773        1761  Female                                  2.0   
175067       14217    Male                                  3.0   
92485         1054    Male                                 12.0   
46002         5197    Male                                  2.0   
6879         28241  Female                                 12.0   
194370        8344    Male                                  0.0   
11175         8382    Male                                  3.0   
173285       11416  Female                                  6.0   
190655       13809    Male                                 10.0   
16545         7481  Female                                 21.0   
21067         6412    Male                                  8.0   
165866       14706    Male                                  2.0   
938           9546    Male                                  2.0   
170154       20014    Male                                  1.0   
80804        24350  Female                                  6.0   
172327       12379  Female                                  4.0   
11169         9653    Male                                  6.0   
90688         5415  Female                                  3.0   
22606          730  Female                                 14.0   
1916         27754    Male                                 18.0   
19962         7474  Female                                  6.0   
22222         4534  Female                                 20.0   
116085         126    Male                                  3.0   
11357        17196  Female                                  7.0   
...            ...     ...                                  ...   
23028        23218    Male                                  6.0   
60750        12604    Male                                  4.0   
105139        2674  Female                                  8.0   
179024       20452  Female                                  7.0   
22522         1699  Female                                 12.0   
169309       27851    Male                                  6.0   
63303         4429  Female                                  0.0   
101521       16660    Male                                 10.0   
175531       18413  Female                                  3.0   
89007         9478    Male                                  4.0   
65772         3478  Female                                  6.0   
3090         23762  Female                                 10.0   
173865       28630    Male                                  5.0   
55730        19643  Female                                 15.0   
65233        12173  Female                                 12.0   
184039       11264  Female                                 14.0   
58439        15355  Female                                 24.0   
178579       13230  Female                                 30.0   
19644         2118  Female                                  7.0   
51736        23545    Male                                  4.0   
115223         165  Female                                  4.0   
142182        8254  Female                                  6.0   
184460        7616  Female                                  5.0   
52279        17965  Female                                  2.0   
68916          125  Female                                  3.0   
165514       19695    Male                                 14.0   
19051         6911    Male                                 12.0   
13414        11054  Female                             

In [56]:
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [57]:
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    19289
<ISCED level 5>    14586
<ISCED level 4>     5106
<ISCED level 2>     3612
<ISCED level 6>     2630
<ISCED level 1>      533
<ISCED level 0>      151
Name: Education - Mother, dtype: int64

In [58]:
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [59]:
mother_edu.sample(10)

Student ID  Gender  Out-of-School Study Time - Homework  \
187126       10407  Female                                  2.0   
176159        7148  Female                                  5.0   
97056        21581    Male                                 20.0   
145594       29703    Male                                  5.0   
59777         2255  Female                                  2.0   
182278        1366  Female                                  6.0   
169939       13213    Male                                  2.0   
49829        18984    Male                                 10.0   
188104       19123    Male                                  4.0   
93929         1992    Male                                  3.0   

        Out-of-School Study Time - Guided Homework  \
187126                                         1.0   
176159                                         2.0   
97056                                          0.0   
145594                                         0.0   
59777                                          1.0   
182278                                         4.0   
169939                                         2.0   
49829                                          0.0   
188104                                         1.0   
93929                                          1.0   

        Out-of-School Study Time - Personal Tutor  \
187126                                        0.0   
176159                                        0.0   
97056                                         3.0   
145594                                        0.0   
59777                                         0.0   
182278                                        3.0   
169939                                        0.0   
49829                                         0.0   
188104                                        0.0   
93929                                         0.0   

        Out-of-School Study Time - Commercial Company  \
187126                                            0.0   
176159                                            0.0   
97056                                             3.0   
145594                                            0.0   
59777                                             0.0   
182278                                            0.0   
169939                                            0.0   
49829                                             0.0   
188104                                            0.0   
93929                                             0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
187126                                     1.0                        250.0   
176159                                     2.0                        200.0   
97056                                      0.0                        200.0   
145594                                     0.0                        275.0   
59777                                      1.0                        180.0   
182278                                     1.0                        180.0   
169939                                     0.0                        200.0   
49829                                      0.0                        220.0   
188104                                     2.0                        200.0   
93929                                      2.0                        180.0   

        Learning Time - Test Language  Learning Time - Science  \
187126                          200.0                    150.0   
176159                          200.0                     80.0   
97056                           150.0                    150.0   
145594                          220.0                    440.0   
59777                           180.0                    360.0   
182278                          135.0                    180.0   
169939                          200.0                    100.0   
49829                           165.0                    165.0   

In [60]:
mother_edu['Student ID'].duplicated().sum()

22322

In [61]:
# mother_edu.replace(["NaN"], np.nan, inplace = True)
# mother_edu.dropna(inplace = True)

In [62]:
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')

mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)
mother_edu_clean = mother_edu.drop_duplicates(subset='Student ID', keep='last')

In [63]:
mother_edu['Student ID'].duplicated().sum()

22322

In [64]:
mother_edu.shape

(45907, 16)

In [65]:
mother_edu_clean.shape

(23585, 16)

In [66]:
mother_edu_clean.duplicated().sum()

0

In [67]:
mother_edu_clean.sample(10)

Student ID  Gender  Out-of-School Study Time - Homework  \
173024       13979    Male                                  1.0   
185109        6909  Female                                  3.0   
169266       26077  Female                                 10.0   
191267        7542  Female                                  5.0   
61279        10184    Male                                 10.0   
145415       28920  Female                                  3.0   
58562        17835  Female                                  1.0   
45885         2869    Male                                 30.0   
168049       23352    Male                                  1.0   
61282        10246  Female                                  9.0   

        Out-of-School Study Time - Guided Homework  \
173024                                         0.0   
185109                                         2.0   
169266                                         3.0   
191267                                         0.0   
61279                                         10.0   
145415                                         0.0   
58562                                          1.0   
45885                                          0.0   
168049                                         0.0   
61282                                          1.0   

        Out-of-School Study Time - Personal Tutor  \
173024                                        0.0   
185109                                        0.0   
169266                                        2.0   
191267                                        4.0   
61279                                         0.0   
145415                                        1.0   
58562                                         0.0   
45885                                         1.0   
168049                                        0.0   
61282                                         2.0   

        Out-of-School Study Time - Commercial Company  \
173024                                            0.0   
185109                                            1.0   
169266                                            0.0   
191267                                            0.0   
61279                                             0.0   
145415                                            0.0   
58562                                             0.0   
45885                                             0.0   
168049                                            0.0   
61282                                             0.0   

        Out-of-School Study Time - With Parent  Learning Time - Mathematics  \
173024                                     0.0                        250.0   
185109                                     0.0                        150.0   
169266                                     0.0                        180.0   
191267                                     0.0                        180.0   
61279                                      8.0                        220.0   
145415                                     0.0                        180.0   
58562                                      1.0                        375.0   
45885                                      0.0                        220.0   
168049                                     0.0                        250.0   
61282                                      0.0                        180.0   

        Learning Time - Test Language  Learning Time - Science  \
173024                          250.0                    250.0   
185109                          150.0                    150.0   
169266                          600.0                    120.0   
191267                          180.0                    180.0   
61279                           220.0                    110.0   
145415                          240.0                    120.0   
58562                           375.0                    375.0   
45885                           200.0                    350.0   

In [68]:
pisa_clean = mother_edu_clean.copy()

In [69]:
pisa_clean.shape

(23585, 16)

In [70]:
pisa_clean.duplicated().sum()

0

In [71]:
pisa_clean = mother_edu.copy()

In [72]:
pisa_clean.shape

(45907, 16)

### 4. Missing values in relevant columns 
('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [73]:

# pisa_clean.replace(["NaN"], np.nan, inplace = True)
# pisa_clean.dropna(inplace = True)

In [74]:
# Check to see if null values correctly dropped
pisa_clean

Student ID  Gender  Out-of-School Study Time - Homework  \
235845        1647    Male                                  2.0   
212618       18805  Female                                  0.0   
212616       18913    Male                                  2.0   
212606       12822    Male                                 30.0   
212595        9143    Male                                  4.0   
212592       10479    Male                                  2.0   
212583       16295  Female                                  1.0   
212580       16142  Female                                  5.0   
212577       17761    Male                                  1.0   
212576       17846    Male                                  2.0   
212571       17414  Female                                  2.0   
212563       14327  Female                                  1.0   
212554       13635    Male                                  2.0   
212546       15437  Female                                  8.0   
212534       15337    Male                                  2.0   
212510       21333    Male                                  7.0   
212635       14280    Male                                  3.0   
212401       29985    Male                                  0.0   
212668       19354  Female                                  5.0   
212771       22336    Male                                  1.0   
213577       22804  Female                                  5.0   
213433       24651  Female                                  7.0   
213049       12825    Male                                  2.0   
213047       13375    Male                                  1.0   
213044       13203  Female                                 10.0   
213029       13900    Male                                  3.0   
212952       11498    Male                                  7.0   
212950       10904  Female                                  9.0   
212876       10648  Female                                  5.0   
212867       25210  Female                                 10.0   
...            ...     ...                                  ...   
42908        14502  Female                                 11.0   
42911        14461    Male                                 15.0   
42912        14415    Male                                 10.0   
42913        14413  Female                                 12.0   
42914        14509  Female                                  3.0   
42916        14793    Male                                  3.0   
42917        15089  Female                                 10.0   
42919        15082    Male                                 10.0   
42892        15992  Female                                  2.0   
42890        16106    Male                                  4.0   
42889        16146  Female                                  5.0   
42885        16256    Male                                  5.0   
42852        12828    Male                                  6.0   
42853        12566    Male                                 28.0   
42855        12785  Female                                  1.0   
42856        12768  Female                                  3.0   
42859        12665    Male                                 16.0   
42860        12624  Female                                  0.0   
42861        12804    Male                                  3.0   
42845        14395  Female                                  1.0   
42863        14401    Male                                  5.0   
42868        15545  Female                                  8.0   
42869        15544    Male                                  8.0   
42871        15718  Female                                 10.0   
42872        15537    Male                                  0.0   
42873        15456  Female                                  7.0   
42880        15785  Female                                 14.0   
42884        16271    Male                             

In [75]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(45907, 16)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [76]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning Time - Total'] = pisa_clean['Learning Time - Mathematics'] + pisa_clean['Learning Time - Test Language'] + pisa_clean['Learning Time - Science'] 
                                                                                                                 

In [77]:
list(pisa_clean)

['Student ID',
 'Gender',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning Time - Mathematics',
 'Learning Time - Test Language',
 'Learning Time - Science',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Education - Father',
 'Education - Mother',
 'Out-of-School Study Time - Total',
 'Learning Time - Total']

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [78]:
pisa_clean

Student ID  Gender  Out-of-School Study Time - Homework  \
235845        1647    Male                                  2.0   
212618       18805  Female                                  0.0   
212616       18913    Male                                  2.0   
212606       12822    Male                                 30.0   
212595        9143    Male                                  4.0   
212592       10479    Male                                  2.0   
212583       16295  Female                                  1.0   
212580       16142  Female                                  5.0   
212577       17761    Male                                  1.0   
212576       17846    Male                                  2.0   
212571       17414  Female                                  2.0   
212563       14327  Female                                  1.0   
212554       13635    Male                                  2.0   
212546       15437  Female                                  8.0   
212534       15337    Male                                  2.0   
212510       21333    Male                                  7.0   
212635       14280    Male                                  3.0   
212401       29985    Male                                  0.0   
212668       19354  Female                                  5.0   
212771       22336    Male                                  1.0   
213577       22804  Female                                  5.0   
213433       24651  Female                                  7.0   
213049       12825    Male                                  2.0   
213047       13375    Male                                  1.0   
213044       13203  Female                                 10.0   
213029       13900    Male                                  3.0   
212952       11498    Male                                  7.0   
212950       10904  Female                                  9.0   
212876       10648  Female                                  5.0   
212867       25210  Female                                 10.0   
...            ...     ...                                  ...   
42908        14502  Female                                 11.0   
42911        14461    Male                                 15.0   
42912        14415    Male                                 10.0   
42913        14413  Female                                 12.0   
42914        14509  Female                                  3.0   
42916        14793    Male                                  3.0   
42917        15089  Female                                 10.0   
42919        15082    Male                                 10.0   
42892        15992  Female                                  2.0   
42890        16106    Male                                  4.0   
42889        16146  Female                                  5.0   
42885        16256    Male                                  5.0   
42852        12828    Male                                  6.0   
42853        12566    Male                                 28.0   
42855        12785  Female                                  1.0   
42856        12768  Female                                  3.0   
42859        12665    Male                                 16.0   
42860        12624  Female                                  0.0   
42861        12804    Male                                  3.0   
42845        14395  Female                                  1.0   
42863        14401    Male                                  5.0   
42868        15545  Female                                  8.0   
42869        15544    Male                                  8.0   
42871        15718  Female                                 10.0   
42872        15537    Male                                  0.0   
42873        15456  Female                                  7.0   
42880        15785  Female                                 14.0   
42884        16271    Male                             

In [79]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45907 entries, 235845 to 23640
Data columns (total 18 columns):
Student ID                                       45907 non-null int64
Gender                                           45907 non-null object
Out-of-School Study Time - Homework              45907 non-null float64
Out-of-School Study Time - Guided Homework       45907 non-null float64
Out-of-School Study Time - Personal Tutor        45907 non-null float64
Out-of-School Study Time - Commercial Company    45907 non-null float64
Out-of-School Study Time - With Parent           45907 non-null float64
Learning Time - Mathematics                      45907 non-null float64
Learning Time - Test Language                    45907 non-null float64
Learning Time - Science                          45907 non-null float64
Average Math Score                               45907 non-null float64
Average Reading Score                            45907 non-null float64
Average Science Score    

In [80]:
pisa_clean.shape

(45907, 18)

In [92]:
pisa_clean.drop_duplicates(keep='last', inplace=True)

In [93]:
pisa_clean.duplicated().sum()

0

In [94]:
pisa_clean.shape

(43715, 18)

<a id='sav'></a>
## Store

In [95]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References